tesseract/tesseract_planning/tesseract_process_planners/include/tesseract_process_planning/process_definition.h

```diff
@@ -0,0 +1,162 @@
+};
+
+/**
+ * @class tesseract_process_planners::ProcessDefinition
+ * @details A process definition.
+ *
+ * This is not intended to support or handle all processes but should covers most of the common ones
+ * like sanding, painting, grinding, etc.
+ *
+ * In a process, we assume they follow this pattern.
+ *    * The robot starts from a start position,
+ *    * Moves to an approach position just above the object
+ *    * Executes a process(sanding, painting, etc.)
+ *    * Return to the home position
+ *
+ * Given the process described the user is only required to define two objects. The start position of
+ * the robot and a vector of Process Segment Definitions.
+ */
+struct ProcessDefinition
+{
+  tesseract_planners::WaypointPtr start;             /**< The start position of the robot */
+  std::vector<ProcessSegmentDefinition> segments;              /**< All of the raster segments with approaches and departures */
+  std::vector<ProcessTransitionDefinition> transitions; 	/**< All of the transition to/from a given segment. Must be same length as segments */
+};
+
+/**definition
+ * @class tesseract_process_planners::ProcessTransitionGenerator
+ */
+class ProcessTransitionGenerator
+{
+public:
+  virtual ~ProcessTransitionGenerator() = default;
+  virtual std::vector<tesseract_planners::WaypointPtr> generate(const tesseract_planners::WaypointPtr& start_waypoint,
+                                                                           const tesseract_planners::WaypointPtr& end_waypoint) const = 0;
+};
+typedef std::shared_ptr<ProcessTransitionGenerator> ProcessTransitionGeneratorPtr;
+typedef std::shared_ptr<const ProcessTransitionGenerator> ProcessTransitionGeneratorConstPtr;
+
+
+/**
+ * @class tesseract_process_planners::ProcessDefinitionConfig
+ * @details The Process Definition Config
+ *
+ * This provides the high level process configuration information. It requires the user to provide
+ * the home position waypoint (JointWaypoint) and a set of tool paths (strokes) that should be
+ * converted into a process results definition leveraging both this configuration information and
+ * the ProcessSegmentDefinitions.
+ *
+ * Also, other operations that are nice to have is the ability to offset the process. Particularly useful
+ * when wanting to verify the process without making contact with a surface in the case of sanding.
+ */
+struct ProcessDefinitionConfig
+{
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  tesseract_planners::WaypointPtr start;
+  std::vector<std::vector<tesseract_planners::WaypointPtr>> tool_paths;
+
+  std::vector<ProcessTransitionGeneratorConstPtr> transition_generator;
+
+  Eigen::Isometry3d local_offset_direction;
+  Eigen::Isometry3d world_offset_direction;
+
+  ProcessDefinitionConfig()
+  {
+    local_offset_direction.setIdentity();
+    world_offset_direction.setIdentity();
+  }
+};
+
+
+/**
+ * @brief The ProcessStepGenerator class
+ */
+class ProcessStepGenerator
```

Comment:
What does this do?

Reply:
I need to go through and add more documentation on this once it is merged. This is a base class for process step generators. A process step generator takes a tool path segment and converts it in a process tool path. Currently we assume that each tool path segment has three process steps Approach, Process, and Departure. Now we have have talked about modifying the Approach and Departure to allow a vector of generators. Example case is using noether we generate consistent tool paths on the surface but this needs to be modified based on the process that will using that tool path. In the case of sanding maybe you do not want to be normal to the surface but at an angle of attack. This library would take the tool path segments and generate a new tool path segment that has modified the poses to accommodate the angle of attack. This is to bridget the gap between noether and the planners so neither of these library need to know anything about the particular process.

Reply:
Ok. That makes sense.

Reply:
I just added some documentation from this class mainly based on @Levi-Armstrong description.